src/oscar/apps/communication/utils.py

```diff
@@ -0,0 +1,175 @@
+import logging
+
+from django.conf import settings
+from django.contrib.sites.models import Site
+from django.core.mail import EmailMessage, EmailMultiAlternatives
+
+from oscar.core.loading import get_model
+
+
+CommunicationEventType = get_model('communication', 'CommunicationEventType')
+Email = get_model('communication', 'Email')
+Notification = get_model('communication', 'Notification')
+
+
+class Dispatcher(object):
+
+    def __init__(self, logger=None, mail_connection=None):
+        if not logger:
+            logger = logging.getLogger(__name__)
+        self.logger = logger
+        # Supply a mail_connection if you want the dispatcher to use that
+        # instead of opening a new one.
+        self.mail_connection = mail_connection
+
+    # Public API methods
+
+    def dispatch_direct_messages(self, recipient_email, messages, attachments=None):
+        """
+        Dispatch one-off messages to explicitly specified recipient email.
+        """
+        if messages['subject'] and (messages['body'] or messages['html']):
+            return self.send_email_messages(recipient_email, messages, attachments=attachments)
+
+    def dispatch_anonymous_messages(self, email, messages, attachments=None):
+        dispatched_messages = {}
+        if email:
+            dispatched_messages['email'] = self.send_email_messages(email, messages, attachments=attachments), None
+        return dispatched_messages
+
+    def dispatch_user_messages(self, user, messages, attachments=None):
+        """
+        Send messages to a site user
+        """
+        dispatched_messages = {}
+        if messages['subject'] and (messages['body'] or messages['html']):
+            dispatched_messages['email'] = self.send_user_email_messages(user, messages, attachments)
+        if messages['sms']:
+            dispatched_messages['sms'] = self.send_text_message(user, messages['sms'])
+        return dispatched_messages
+
+    def notify_user(self, user, subject, **kwargs):
+        """
+        Send a simple notification to a user
+        """
+        Notification.objects.create(recipient=user, subject=subject, **kwargs)
+
+    def notify_users(self, users, subject, **kwargs):
+        """
+        Send a simple notification to an iterable of users
+        """
+        for user in users:
+            self.notify_user(user, subject, **kwargs)
+
+    # Internal
+
+    def create_email(self, user, messages, email):
+        """
+        Create ``Email`` instance in database for logging purposes.
+        """
+        if email and user.is_authenticated:
+            return Email._default_manager.create(
```

Comment:
Why are we using `_default_manager` here? To facilitate easy overriding I think we should be using `Email.objects`.